linker/shared/src/main/scala/org/scalajs/linker/backend/OutputWriter.scala

```diff
@@ -30,21 +30,29 @@ private[backend] abstract class OutputWriter(output: OutputDirectory,
   private val outputImpl = OutputDirectoryImpl.fromOutputDirectory(output)
   private val moduleKind = config.commonConfig.coreSpec.moduleKind
 
-  protected def writeModule(moduleID: ModuleID, jsFileWriter: ByteArrayWriter): Unit
+  protected type PreparedModule
 
-  protected def writeModule(moduleID: ModuleID, jsFileWriter: ByteArrayWriter,
-      sourceMapWriter: ByteArrayWriter): Unit
+  protected def prepareModule(moduleID: ModuleID): PreparedModule
+
+  protected def didModuleChange(moduleID: ModuleID, prepared: PreparedModule): Boolean
+
+  protected def writeModule(moduleID: ModuleID, prepared: PreparedModule,
+      jsFileWriter: ByteArrayWriter): Unit
+
+  protected def writeModule(moduleID: ModuleID, prepared: PreparedModule,
+      jsFileWriter: ByteArrayWriter, sourceMapWriter: ByteArrayWriter): Unit
```

Comment:
Rather than `prepareModule` and `didModuleChange`, have you considered:

```
protected def writeModule(moduleID: ModuleID, force: Boolean): Option[ByteBuffer]

protected def writeModuleWithSourceMap(moduleID: ModuleID, force: Boolean): (Option[ByteBuffer], Option[ByteBuffer])
// Or Option[(ByteBuffer, ByteBuffer)] ?
```

The downside is that there's an unchecked invariant that the return value must be a `Some` if `force = true`.

This will move the allocation of `ByteArrayWriter` to the call site, but that's probably even a good thing. Notably, when we look at writing the JS file in the context of #4820, we're only `"use strict;"` away from concatenating `Array[Bytes]`. So it might make sense to just allocate a `ByteBuffer` with the target size and fill it.